pkg/api/types.go

```diff
@@ -688,36 +691,66 @@ type ServiceStatus struct{}
+	Ports []ServicePort
 
-	// Optional: Supports "TCP" and "UDP".  Defaults to "TCP".
-	Protocol Protocol `json:"protocol,omitempty"`
+	// Either 'Selector' or 'Endpoints' must be set:
 
-	// This service will route traffic to pods having labels matching this selector. If empty or not present,
-	// the service is assumed to have endpoints set by an external process and Kubernetes will not modify
-	// those endpoints.
+	// This service will route traffic to pods having labels matching this
+	// selector.
 	Selector map[string]string `json:"selector"`
 
+	// Use the specified endpoints.
+	// FIXME: Some weird interaction.  All the fields of Ports[] are basically
+	//   ignored in favor of Endpoints.  Maybe make this just a list of IPs,
+	//   and disallow named DestinationPorts if using this mode?
+	Endpoints []Endpoint
+
 	// PortalIP is usually assigned by the master.  If specified by the user
 	// we will try to respect it or else fail the request.  This field can
 	// not be changed by updates.
 	PortalIP string `json:"portalIP,omitempty"`
 
-	// CreateExternalLoadBalancer indicates whether a load balancer should be created for this service.
+	//FIXME: Should these next 3 be per-port (different external IPs?) or for all ports?
+	// CreateExternalLoadBalancer indicates whether a load balancer should
+	// be created for this service.
 	CreateExternalLoadBalancer bool `json:"createExternalLoadBalancer,omitempty"`
+
 	// PublicIPs are used by external load balancers.
 	PublicIPs []string `json:"publicIPs,omitempty"`
 
-	// ContainerPort is the name of the port on the container to direct traffic to.
-	// Optional, if unspecified use the first port on the container.
-	ContainerPort util.IntOrString `json:"containerPort,omitempty"`
-
-	// Optional: Supports "ClientIP" and "None".  Used to maintain session affinity.
+	// Optional: Supports "ClientIP" and "None".  Used to maintain session
+	// affinity.
 	SessionAffinity AffinityType `json:"sessionAffinity,omitempty"`
 }
 
+type ServicePort struct {
+	// Optional if only one ServicePort is defined on this service,
+	// otherwise required: The name of this port within the service.  This
+	// must be a DNS_LABEL.  All ports within a ServiceSpec must have
+	// unique roles.  This maps to the 'Role' field in Endpoint objects.
+	Role string
+
+	// Optional: The IP protocol for this port.  Supports "TCP" and "UDP".
+	// Defaults to "TCP".
+	Protocol Protocol
+
+	// Required: The port that will be exposed by this service.
+	Port int
+
+	// Optional: How many contiguous ports to expose.  If this is not
+	// specified, defaults to 1.
+	Count int
+
+	// Optional: The destination port on pods selected by this service.
+	// If this is a string, it will be looked up as a named port in the
+	// target Pod's container ports.  If this is not specified, the value of
+	// 'Port' is used (an identity map).
+	// FIXME: This is a change in semantic in the not-specified case.  Can we get away with it?  is it worthwhile?
```

Comment:
I really hate the order dependency, but would like to make single-port cases very simple.

Is there a problem with forwarding the full port range to the pods?